### PR TITLE
Isolate DBs

### LINKED ISSUE #41: Isolate databases from one-another
[A discussion on the SQLite forum](https://sqlite.org/forum/forumpost/0f77e9f96ec58def) that effectively means we want to isolate databases from one another so that one user can't query/edit another user's database.

[Some general background reading](https://fly.io/blog/sandboxing-and-workload-isolation/)

A few options to explore
* Lightweight WASM: https://docs.rs/wasmtime/latest/wasmtime/
* Good old heavyweight Docker: https://docs.rs/docker-command/latest/docker_command/
* Lighter-weight jails: https://nsjail.dev/. Used in this project: https://github.com/windmill-labs/windmill/tree/main/backend/windmill-worker

### PR DESCRIPTION
**Pull request by marcua**
Resolves #41

### CODE REVIEW

File: Cargo.toml
Comment (sofiaritz) [inline Cargo.toml:14]: Maybe creating a [workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) is a good idea.
There could be three packages: `ayb`, `ayb-hosted-db-runner`, and `ayb-common`.
That way things like AybError can be imported from the `ayb-common` crate. :)

Amazing work!! :)

Comment (marcua) [reply Cargo.toml:14]: Thank you! This is a great recommendation. I'm trying to figure out the right structure for what we need. All we really want is a second binary crate for the runner so that we can wrap it in a `vsjail`. Would it be enough to create two binary crates (`ayb` and `hosted_db_runner`) in the same package and avoiding multiple packages or workspaces entirely, or can we think of some benefit of breaking them apart?

Relatedly, I'm curious if there's a canonical way to reference one binary crate from another so the `ayb` binary can say "run the other binary for the runner" without having to explicitly provide a path to it.

Comment (sofiaritz) [reply Cargo.toml:14]: > Would it be enough to create two binary crates (ayb and hosted_db_runner) in the same package

I have never seen that in the wild tbh. Maybe it's fine, but that's also one of the use cases of workspaces... I would create a top level workspace just to be sure and to also follow the "convention".

This article may be of interest: https://matklad.github.io/2021/08/22/large-rust-workspaces.html

> Relatedly, I'm curious if there's a canonical way to reference one binary crate from another

Not that I am aware of, sadly.
You could embed the runner at compile time with `include_bytes!` and create a copy of the runner at startup time and delete it when the program is stopped, that way the runner is always up-to-date and is also directly tied to the ayb binary.

File: scripts/build_nsjail.sh
Comment (marcua) [inline scripts/build_nsjail.sh:9]: Remove

File: src/hosted_db/sandbox.rs
Comment (marcua) [inline src/hosted_db/sandbox.rs:65]: Add note that this assumes `ayb` and `ayb_isolated_runner` will always be in the same directory

File: Cargo.toml
Comment (marcua) [inline Cargo.toml:38]: I wonder if I need double-brackets

File: src/bin/ayb_isolated_runner.rs
Comment (marcua) [inline src/bin/ayb_isolated_runner.rs:6]: Add comment explaining what this runner is for

File: Cargo.toml
Comment (marcua) [inline Cargo.toml:6]: Rebase/merge main, resolve conflicts

File: Cargo.toml
Comment (marcua) [inline Cargo.toml:8]: Add section on isolation and the optional `nsjail` configuration to `README.md`

File: src/hosted_db/sandbox.rs
Comment (marcua) [inline src/hosted_db/sandbox.rs:41]: Remove question marks here and on next line :)

File: src/hosted_db/sandbox.rs
Comment (marcua) [inline src/hosted_db/sandbox.rs:102]: Presumably we'll regret having hard-coded limits on stdout/stderr, add TODO

File: src/hosted_db/sqlite.rs
Comment (marcua) [inline src/hosted_db/sqlite.rs:51]: Consider this function name relative to the other one in this file: can they be more distinct?

File: Cargo.toml
Comment (marcua) [reply Cargo.toml]: Phew! I managed to figure out a way without too much hackery and one somewhat safe assumption
* One crate
* Two binaries (`ayb` and ayb_isolated_runner`)
* Since it's the same crate, it's relatively safe to assume both binaries will end up in the same directory (true of build targets as well as where the files will be installed with a `crate install`
* If we assume they are in the same directory, we can call `std::env::current_exe` to get a path to the current binary, which we can then use to create a path to the other binary

Hooray! I'm resolving conflicts, but this PR is largely good for review if you want to take a pass over it. I've self-reviewed and left myself comments that I'll resolve in the coming days, but if you're interested I'd love your opinion! :)

File: src/hosted_db/sandbox.rs
Comment (marcua) [inline src/hosted_db/sandbox.rs:59]: Consider how to avoid this repetition of the AybError from line 55.

File: src/http/structs.rs
Comment (marcua) [inline src/http/structs.rs:54]: tiny hack: I've checked in a built `nsjail` binary to `tests/nsjail`. In a perfect world I'd containerize it or build it each time from scratch, both of which feel like a decent amount of work given it's just for tests and not actually distributing the binary.

File: Cargo.toml
Comment (marcua) [inline Cargo.toml:7]: Tests pass locally but fail in GitHub Actions
image

File: Cargo.toml
Comment (sofiaritz) [reply Cargo.toml]: That sounds like some reasonable assumptions and a clean workaround :)
I still don't really like the idea of one crate having two binaries like that... maybe creating a `[[bin]]` entry is a better approach (https://doc.rust-lang.org/cargo/reference/cargo-targets.html#binaries)

I'll review the PR :)

File: src/hosted_db/sandbox.rs
Comment (sofiaritz) [inline src/hosted_db/sandbox.rs:91]: Is piping `stdin` required?

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [inline src/hosted_db/sqlite.rs:60]:
```suggestion
            if !result.stderr.is_empty() {
```

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [inline src/hosted_db/sqlite.rs:70]:
```suggestion
            } else if !result.stdout.is_empty() {
```

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [inline src/hosted_db/sqlite.rs:56]: Most of the checks here can be replaced with early returns, which would make the code more readable. :)

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [inline src/hosted_db/sqlite.rs:79]: You could remove a level of nesting by doing the following at the top
```rust
if isolation.is_none() {
   return Ok(query_sqlite(path, query)?.into());
}
```

**Review by sofiaritz — CHANGES_REQUESTED** [review]
I like how you ended up implementing this, a clean and simple implementation :)
Now that the architecture of ayb is getting a little bit more complex, maybe it's time to create something like an [ARCHITECTURE.md](https://github.com/rust-lang/rust-analyzer/blob/d7c99931d05e3723d878bea5dc26766791fa4e69/docs/dev/architecture.md) file.

File: Cargo.toml
Comment (marcua) [reply Cargo.toml:41]: woohoo! learned something new! :)

File: scripts/build_nsjail.sh
Comment (marcua) [reply scripts/build_nsjail.sh:9]: Done

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:65]: Done

File: src/bin/ayb_isolated_runner.rs
Comment (marcua) [reply src/bin/ayb_isolated_runner.rs:6]: Done

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:4]: Good call, done! :)

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:34]: Interestingly, initially I tried importing Sandkasten so I could use its abstraction, but that broke down as that library wasn't designed as an abstraction for other libraries. So I copied in the parts that I needed here (thus the copyright notice), and realized that the amount of abstraction was too high for something `ayb` would have one concrete use case for. The abstracted code was more difficult to follow, since `ayb` isn't `nsjail-wrapper`, and so I went with a more concrete single use case here.

I generally avoid abstractions that have only one concrete instance since I can't vet the abstraction without multiple examples. My bias is to follow that general avoidance in this case, but if there's some specific benefit you can think of that outweighs the cost of indirection, I'm open to hearing it! :)

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:41]: Done!

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:44]: Perfect! Done!

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:59]: Done!

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:91]: Good eye! :) I was able to remove it with no effect.

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:102]: This was originally a limit option to the Sandkasten implementation, and I figured providing a generous default was reasonable (the query would time out at this size of return anyway). That said, I was able to remove the `take` and make the readers `mut` and it seems to work just fine. We can always make this adjustable later. Good push to simplify this! :)

File: Cargo.toml
Comment (marcua) [reply Cargo.toml:8]: Done

File: src/http/structs.rs
Comment (marcua) [reply src/http/structs.rs:54]: Good push --- it was more laziness for not wanting to update GitHub actions to build `nsjail` :P.

We now have
* just the script
* GH actions builds `nsjail`
* documentation in README.md on isolation (yay!) and how to run tests (because you now have to build `nsjail` once, yay!)

File: Cargo.toml
Comment (marcua) [reply Cargo.toml:7]: Building `nsjail` solved it!!! :)

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs:60]: Nice! Thank you!

(I noticed `cargo clippy` also caught this, and realized clippy isn't breaking the build, so I created #250)

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs:70]: Done! Thank you!

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs:51]: Done!

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs:56]: Thank you for this feedback. Aside from the `match`, which we'll discuss below, are there others to consider?

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs:79]: I'm conflicted on this one. We'd get one level of nesting in a relatively short function, but in exchange
* I would stop using a language feature like `match`, which I feel is being used appropriately
* I'd have to introduce an `unwrap`, since the compiler won't be able to reason about the knowledge we'd gain with the `if`

Given we're using match in the way it was intended and forcing ourselves to enumerate all of the variants, I wonder if we'd gain much in exchange for reducing indentation in a short function.

File: src/http/config.rs
Comment (marcua) [reply src/http/config.rs:36]: Very nice! Done!

**Comment by marcua** [comment]
Thank you for the very helpful review @sofiaritz! You made the implementation an documentation far more approachable, and I appreciate your feedback. Looking forward to your next round!

Re: ARCHITECTURE.md, great idea. I added basic documentation on isolation and testing, but also created #252 for a deeper pass.

File: src/hosted_db/sandbox.rs
Comment (sofiaritz) [reply src/hosted_db/sandbox.rs:56]: You're right, that abstraction wouldn't make much sense in our case.
I think the comments are enough, I'll mark this as resolved :)

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [reply src/hosted_db/sqlite.rs]: This is just a style preference, the `Option` enum won't add more variants and thus `match` doesn't introduce any advantages over other ways of error checking/handling.

I feel that you are _really really_ skeptical about using `unwrap`/`expect`. Let's not forget that in our case when something panics inside a request to the server, the ayb server keeps running. In our case, using `unwrap`/`expect` in a way that can _actually_ panic isn't as bad as using it in a library or a TUI (not saying that that's the case with this suggestion, this is just a general comment).

I like early returns because they remove mental overhead from the developer and clearly separate the "invalid cases"/error handling from the actual logic of the code.

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [reply src/hosted_db/sqlite.rs:58]: Nope :)

**Comment by sofiaritz** [comment]
Done! Just a few comments about early returns and `match`. Really excited for this to land! :)

File: src/hosted_db/sqlite.rs
Comment (sofiaritz) [reply src/hosted_db/sqlite.rs]: New idea! What about using `if let Some(isolation) = isolation` and returning the `None` case at the end of the function? Seems like a reasonable middleground :)

```rust
pub async fn potentially_isolated_sqlite_query(
    path: &PathBuf,
    query: &str,
    isolation: &Option<AybConfigIsolation>,
) -> Result<QueryResult, AybError> {
    if let Some(isolation) = isolation {
        let result = run_in_sandbox(Path::new(&isolation.nsjail_path), path, query).await?;
        if !result.stderr.is_empty() {
            let error: AybError = serde_json::from_str(&result.stderr)?;
            return Err(error);
        }
        
        if result.status != 0 {
            return Err(AybError::Other {
                message: format!(
                    "Error status from sandboxed query runner: {}",
                    result.status
                ),
            });
        }
        
        if !result.stdout.is_empty() {
            let query_result: QueryResult = serde_json::from_str(&result.stdout)?;
            return Ok(query_result);
        }
        
        return Err(AybError::Other {
            message: "No results from sandboxed query runner".to_string(),
        });
    }
    
    Ok(query_sqlite(path, query)?)
}
```

File: src/hosted_db/sandbox.rs
Comment (marcua) [reply src/hosted_db/sandbox.rs:56]: Hooray! Thank you!

File: src/hosted_db/sqlite.rs
Comment (marcua) [reply src/hosted_db/sqlite.rs]: Clean! I like it! I appreciated both of your bullet points on the broader point. I'll implement this and we should be good to go! Woohoo! Thank you!!!

**Review by sofiaritz — APPROVED** [review]
Amazing work!! When this lands, I'll update aybDocker and start hosting a public instance :)

**Comment by marcua** [comment]
Thank you for the very helpful review, @sofiaritz! I'll merge now, and after #237 lands, I'll release the next version for your public instance! Woohooo! :)

**Pull request merged**
